packages/compass-app-stores/src/stores/instance-store.ts

```diff
@@ -0,0 +1,669 @@
+    const { instance, dataService } = store.getState();
+
+    if (!instance || !dataService) {
+      debug(
+        'Trying to fetch collections without the model or dataService in the state'
+      );
+      return;
+    }
+
+    // It is possible to get here before the databases finished loading. We have
+    // to wait for the databases, otherwise it will load all the collections for 0
+    // databases.
+    await instance.fetchDatabases({ dataService });
+
+    await Promise.all(
+      instance.databases.map((db) => {
+        return db.fetchCollections({ dataService });
+      })
+    );
+  }
+
+  /**
+   * Fetches collection info and returns a special format of collection metadata
+   * that events like open-in-new-tab, select-namespace, edit-view require
+   */
+  async function fetchCollectionMetadata(ns: string) {
+    const { instance, dataService } = store.getState();
+    const { database, collection } = toNS(ns);
+
+    if (!instance || !dataService) {
+      debug(
+        'Trying to fetch collection metadata without the model or dataService in the state'
+      );
+      return;
+    }
+
+    const coll = await instance.getNamespace({
+      database,
+      collection,
+      dataService,
+    });
+
+    return await coll?.fetchMetadata({ dataService });
+  }
+
+  async function refreshNamespace({
+    ns,
+    database,
+  }: {
+    ns: string;
+    database: string;
+  }) {
+    const instance = getInstance();
+    if (!instance.databases.get(database)) {
+      await instance.fetchDatabases({ dataService, force: true });
+    }
+    const db = instance.databases.get(database);
+    if (!db?.collections.get(ns)) {
+      await db?.fetchCollections({ dataService, force: true });
+    }
+    await refreshNamespaceStats(ns);
+  }
+
+  async function refreshNamespaceStats(ns: string) {
+    const instance = getInstance();
+    const { database } = toNS(ns);
+    const db = instance.databases.get(database);
+    const coll = db?.collections.get(ns);
+    // We don't care if this fails
+    await Promise.allSettled([
+      db?.fetch({ dataService, force: true }),
+      coll?.fetch({ dataService, force: true }),
+    ]);
+  }
+
+  const connectionString = dataService.getConnectionString();
+  const firstHost = connectionString.hosts[0] || '';
+  const [hostname, port] = firstHost.split(':');
+
+  const initialInstanceProps: Partial<MongoDBInstanceProps> = {
+    _id: firstHost,
+    hostname: hostname,
+    port: port ? +port : undefined,
+    topologyDescription: getTopologyDescription(
+      dataService.getLastSeenTopology()
+    ),
+  };
+  const instance = new MongoDBInstance(
+    initialInstanceProps as MongoDBInstanceProps
+  );
+  if ((globalThis as any).hadronApp) {
+    (globalThis as any).hadronApp.instance = instance;
+  }
+
+  debug('instance-created');
+  appRegistry.emit('instance-created', { instance });
+
+  store.dispatch(changeDataService(dataService));
+  store.dispatch(changeInstance(instance));
```

Comment:
You can use `createStore` second argument to provide those when store is created instead of dispatching events in the method here, this is a bit more idiomatic way of setting it up instead of dispatching events right after creating a store:

```suggestion
  const store = createStore(reducer, { dataService, instance });
```